### PR TITLE
added `modifier` hook to `StacApiIO` to enable AWS SigV4 signing

### PR DESCRIPTION
**Pull request by greyskyy**
**Related Issue(s):** 

- #371 

**Description:**

A modifier provides the ability to modify the request immediately before sending, a requirement for AWS SigV4. This allows users to plug in their signing method.

Also, added the `stac_io` parameter to `Client.open()` allowing for easy usage of a custom `StacApiIO` instance.

Using this PR, an application could access a private stac-server protected behind AWS IAM authentication as follows:

```python
import boto3
import botocore.auth
import botocore.awsrequest
import pystac_client
import requests

region = "us-east-1"
service_name = "execute-api"
endpoint_id = "xxxxxxxx"
deployment_stage = "dev"

# load AWS credentials
credentials = boto3.Session(region_name=region).get_credentials()
signer = botocore.auth.SigV4Auth(credentials, service_name, region)

def sign_request(request: requests.Request) -> requests.Request:
    aws_request = botocore.awsrequest.AWSRequest(
        method=request.method,
        url=request.url,
        params=request.params,
        data=request.data,
        headers=request.headers
    )
    signer.add_auth(aws_request)
    request.headers = aws_request.headers
    return request

client = pystac_client.Client.open(
    url=f"https://{endpoint_id}.{service_name}.{region}.amazonaws.com/{deployment_stage}",
    request_modifier=sign_request
)

for collection in client.get_all_collections():
    print(collection.id)
```

**PR Checklist:**

- [x] Code is formatted
- [x] Tests pass
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac-api-client/blob/main/CHANGELOG.md)

### CODE REVIEW

File: pystac_client/stac_api_io.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def __init__(
         headers: Optional[Dict[str, str]] = None,
         conformance: Optional[List[str]] = None,
         parameters: Optional[Dict[str, Any]] = None,
+        modifier: Optional[Callable[[Any], Any]] = None,
```
Comment (gadomski) [inline pystac_client/stac_api_io.py:37]: If the modifier should receive and return `request.Request`s, shouldn't we call that out in the type signature?

**Review by gadomski — CHANGES_REQUESTED** [review]
In general, looks good. One question inline, and one top-level ask: could you include your example usage from https://github.com/stac-utils/pystac-client/pull/372#issue-1480019666 in the documentation? Even if we don't add software support for AWS SigV4, having a documentation example might save us some questions down the road.

File: pystac_client/stac_api_io.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Geosynopsis) [reply pystac_client/stac_api_io.py:37]: On the Client's `from_file` method, we instantiate StacApiIO. Would that function require passing this request modifier optionally?

[the remark this answers is not in the document]
Comment (greyskyy) [reply pystac_client/stac_api_io.py]: updated the name to `request_modifier` for clarity.

[the remark this answers is not in the document]
Comment (greyskyy) [reply pystac_client/stac_api_io.py]: @Geosynopsis I had thought about that. However, for simplicity, it seemed better to just create the `StacApiIO` object first and use the existing `stac_io` parameter.  But now that you mention it, I can add it to the `Client.open()` and `Client.from_file()` methods.

**Comment by greyskyy** [comment]
> In general, looks good. One question inline, and one top-level ask: could you include your example usage from [#372 (comment)](https://github.com/stac-utils/pystac-client/pull/372#issue-1480019666) in the documentation? Even if we don't add software support for AWS SigV4, having a documentation example might save us some questions down the road.

@gadomski I added a tutorial for using `pystac-client` with authentication. I included basic auth too, since it shows how you can just update the HTTP headers and seemed a decent second example.  Let me know if you have any other suggestions.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 
 - Python 3.11 support [#347](https://github.com/stac-utils/pystac-client/pull/347)
+- Added `modifier` to `StacApiIO` to allow for additional authentication mechanisms (e.g. AWS SigV4) [#371](https://github.com/stac-utils/pystac-client/issues/371).
```
Comment (Geosynopsis) [inline CHANGELOG.md:13]: Please update to reflect the current change.

Comment (greyskyy) [reply CHANGELOG.md:13]: done.

**Comment by greyskyy** [comment]
> Looks good, thanks for including the documentation. The only issue is that, when I build the docs locally, the tutorial ends up with a bunch of errors due to the dummy endpoint in the docs: Screenshot 2022-12-09 at 07-59-01 Authentication — pystac-client 0 5 1 documentation
> 
> Can we either:
> 
> * Use a real endpoint, or
> * Do some sort of `no_run` configuration to prevent the errors from appearing when building the docs?

@gadomski I changed the authentication tutorial form a jupyter notebook to markdown. It looks like Sphinx won't execute the python in this case.  That should remove the errors in the doc generation, while still keeping decent-looking formatting.

**Pull request merged**